File: src/main/java/gargoyle/netsquares/model/Board.java
package gargoyle.netsquares.model;

import gargoyle.netsquares.model.i.IBoard;
import gargoyle.netsquares.util.Randoms;

import java.awt.*;
import java.io.Serializable;
import java.util.Arrays;

public final class Board implements Serializable, Cloneable, IBoard {
    private static final long serialVersionUID = 8850083089471052287L;
    private Point caret;
    private int[][] cells;
    private int score;
    private int size;

    public Board() {
        super();
    }

    public Board(final int width, final int score) {
        this();
        initSize(width);
        initScore(score);
    }

    @Override
    public void applyMove(final Move move) {
        getCaret().setLocation(getDestination(move));
    }

    @Override
    public IBoard clone() {
        final Board clone;
        clone = new Board();
        clone.size = size;
        clone.score = score;
        clone.cells = cells.clone();
        clone.caret = caret == null ? null : new Point(caret);
        return clone;
    }

    @Override
    public int getBoardSize() {
        return size;
    }

    @Override
    public Point getCaret() {
        return caret;
    }

    @Override
    public int getCellAt(final int x, final int y) {
        try {
            return cells[x][y];
        } catch (final ArrayIndexOutOfBoundsException e) {
            return 0;
        }
    }

    @Override
    public int getCellAtCaret() {
        return caret == null ? 0 : getCellAt(caret.x, caret.y);
    }

    @Override
    public int getCellAtDestination(final Move move) {
        return getCellAtDestination(getCaret(), move);
    }

    @Override
    public int getCellAtDestination(final Point point, final Move move) {
        final Point c = getDestination(point, move);
        return getCellAt(c.x, c.y);
    }

    @Override
    public Point getDestination(final Move move) {
        return getDestination(getCaret(), move);
    }

    @Override
    public Point getDestination(final Point point, final Move move) {
        final Point destination = new Point(point);
        if (move.getDirection() == Direction.DOWN) {
            destination.y += move.getDistance();
        }
        if (move.getDirection() == Direction.LEFT) {
            destination.x -= move.getDistance();
        }
        if (move.getDirection() == Direction.RIGHT) {
            destination.x += move.getDistance();
        }
        if (move.getDirection() == Direction.UP) {
            destination.y -= move.getDistance();
        }
        return destination;
    }

    @Override
    public Move getMoveToPoint(final int x, final int y) {
        final int dx = x - caret.x;
        final int dy = y - caret.y;
        if ((dx > 0) && (dy == 0)) {
            return new Move(Direction.RIGHT, dx);
        }
        if ((dx < 0) && (dy == 0)) {
            return new Move(Direction.LEFT, -dx);
        }
        if ((dx == 0) && (dy > 0)) {
            return new Move(Direction.DOWN, dy);
        }
        if ((dx == 0) && (dy < 0)) {
            return new Move(Direction.UP, -dy);
        }
        return null;
    }

    @Override
    public Move getNearestMove(final Direction direction) {
        int distance = 1;
        Move move;
        while (isMoveIn(move = new Move(direction, distance))) {
            if (isMoveLegal(move, null)) {
                return move;
            }
            distance++;
        }
        return null;
    }

    @Override
    public int getScore() {
        return score;
    }

    @Override
    public void initScore(final int score) {
        this.score = score;
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                while (isOpenAt(x, y)) {
                    final int a = -score;
                    cells[x][y] = Randoms.random(a, score);
                }
            }
        }
    }

    @Override
    public void initSize(final int size) {
        this.size = size;
        cells = new int[size][size];
        caret = new Point(size / 2, size / 2);
    }

    @Override
    public boolean isIn(final int x, final int y) {
        return (x >= 0) && (x < size) && (y >= 0) && (y < size);
    }

    @Override
    public boolean isMoveAvailable(final Direction[] allowedDirections) {
        for (final Direction direction : allowedDirections) {
            int distance = 1;
            Move move;
            while (isMoveIn(move = new Move(direction, distance))) {
                if (isMoveLegal(move, allowedDirections)) {
                    return true;
                }
                distance++;
            }
        }
        return false;
    }

    @Override
    public boolean isMoveIn(final Move move) {
        return isMoveInFrom(getCaret(), move);
    }

    @Override
    public boolean isMoveInFrom(final Point point, final Move move) {
        final Point destination = getDestination(point, move);
        return isIn(destination.x, destination.y);
    }

    @Override
    public boolean isMoveLegal(final Move move, final Direction[] allowedDirections) {
        final Direction direction = move.getDirection();
        if ((allowedDirections == null) || (allowedDirections.length == 0 || (direction == null))
                || Arrays.asList(allowedDirections).contains(direction)) {
            final Point newCaret = getDestination(move);
            return !isOpenAt(newCaret.x, newCaret.y);
        }
        return false;
    }

    @Override
    public boolean isOpenAt(final int x, final int y) {
        return getCellAt(x, y) == 0;
    }

    @Override
    public boolean isOpenAtCaret() {
        return (caret == null) || isOpenAt(caret.x, caret.y);
    }

    @Override
    public boolean isOpen() {
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                if (!isOpenAt(x, y)) {
                    return false;
                }
            }
        }
        return true;
    }

    @Override
    public boolean move(final Direction direction) {
        final Point c = new Point(getCaret());
        do {
            if (direction == Direction.UP) {
                if (c.y <= 0) {
                    return false;
                }
                c.y--;
            }
            if (direction == Direction.DOWN) {
                if (c.y >= (size - 1)) {
                    return false;
                }
                c.y++;
            }
            if (direction == Direction.LEFT) {
                if (c.x <= 0) {
                    return false;
                }
                c.x--;
            }
            if (direction == Direction.RIGHT) {
                if (c.x >= (size - 1)) {
                    return false;
                }
                c.x++;
            }
        } while (getCellAt(c.x, c.y) == 0);
        caret.setLocation(c);
        return true;
    }

    @Override
    public void open() {
        caret = null;
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                cells[x][y] = 0;
            }
        }
    }

    private int openCell(final int x, final int y) {
        final int cell = getCellAt(x, y);
        try {
            cells[x][y] = 0;
        } catch (final ArrayIndexOutOfBoundsException e) {
            return cell;
        }
        return cell;
    }

    @Override
    public int openCellAtCaret() {
        return caret == null ? 0 : openCell(caret.x, caret.y);
    }

    @Override
    public void reinit() {
        initScore(score);
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Board)) return false;

        final Board board = (Board) o;

        if (score != board.score) return false;
        if (size != board.size) return false;
        return Arrays.deepEquals(cells, board.cells);

    }

    @Override
    public int hashCode() {
        int result = Arrays.deepHashCode(cells);
        result = 31 * result + score;
        result = 31 * result + size;
        return result;
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("Board{");
        sb.append("size=").append(size);
        sb.append(", score=").append(score);
        sb.append(", caret=").append(caret);
        sb.append(", cells=").append(Arrays.toString(cells));
        sb.append('}');
        return sb.toString();
    }
}

File: src/main/kotlin/gargoyle/netsquares/model/Board.kt
package gargoyle.netsquares.model

import gargoyle.netsquares.util.Rnd

class Board {
    var caretX: Int
        private set
    var caretY: Int
        private set
    private var cells: Array<IntArray>
    var left: Int
        private set
    var size: Int
        private set
    var weight: Int
        private set

    init {
        cells = NO_CELLS
        size = 0
        weight = 0
        left = 0
        caretX = 0
        caretY = 0
    }

    fun canMoveBy(move: Move): Boolean {
        return move == Move.OPEN || canMoveBy(requireNotNull(move.direction))
    }

    private fun canMoveBy(direction: Direction): Boolean {
        return canMoveBy(direction.deltaX, direction.deltaY)
    }

    private fun canMoveBy(deltaX: Int, deltaY: Int): Boolean {
        return isIn(caretX + deltaX, caretY + deltaY)
    }

    fun canMoveBy(direction: Direction, step: Int): Boolean {
        return canMoveBy(direction.deltaX * step, direction.deltaY * step)
    }

    fun canOpenBy(move: Move): Boolean {
        return move == Move.OPEN && canOpen() || canOpenBy(requireNotNull(move.direction))
    }

    private fun canOpenBy(direction: Direction): Boolean {
        return canOpenBy(direction.deltaX, direction.deltaY)
    }

    fun canOpen(): Boolean {
        return !isOpenAtCaret
    }

    private fun canOpenBy(deltaX: Int, deltaY: Int): Boolean {
        return canMoveBy(deltaX, deltaY) && !isOpenBy(deltaX, deltaY)
    }

    private fun isOpenBy(deltaX: Int, deltaY: Int): Boolean {
        val x = caretX + deltaX
        val y = caretY + deltaY
        return isIn(x, y) && isOpenAt(x, y)
    }

    private fun isIn(x: Int, y: Int): Boolean {
        return x >= 0 && y >= 0 && x < size && y < size
    }

    fun isOpenAt(x: Int, y: Int): Boolean {
        return cells[x][y] == 0
    }

    fun canOpenBy(direction: Direction, step: Int): Boolean {
        return canOpenBy(direction.deltaX * step, direction.deltaY * step)
    }

    fun each(operation: (Int, Int)-> Int?) {
        for (x in 0 until size) {
            for (y in 0 until size) {
                val res = operation(x, y)
                if (res != null) {
                    cells[x][y] = res
                }
            }
        }
    }

    val atCaret: Int
        get() = getAt(caretX, caretY)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val board = other as Board
        return caretX == board.caretX && caretY == board.caretY && left == board.left && size == board.size && weight == board.weight &&
                cells.contentEquals(board.cells)
    }

    override fun toString(): String {
        return String.format(
            "Board{size=%d, weight=%d, caretX=%d, caretY=%d, left=%d, cells=%s}",
            size,
            weight,
            caretX,
            caretY,
            left,
            cells.contentDeepToString()
        )
    }

    private fun getBy(deltaX: Int, deltaY: Int): Int {
        return getAt(caretX + deltaX, caretY + deltaY)
    }

    fun getAt(x: Int, y: Int): Int {
        return if (isIn(x, y)) cells[x][y] else 0
    }

    fun getBy(direction: Direction, step: Int): Int {
        return getBy(direction.deltaX * step, direction.deltaY * step)
    }

    override fun hashCode(): Int {
        var result = arrayOf(caretX, caretY, left, size, weight).contentHashCode()
        result = 31 * result + cells.contentHashCode()
        return result
    }

    fun getBy(move: Move): Int {
        return if (move == Move.OPEN) atCaret else getBy(requireNotNull(move.direction))
    }

    private fun getBy(direction: Direction): Int {
        return getBy(direction.deltaX, direction.deltaY)
    }

    fun init(size: Int, weight: Int) {
        cells = Array(size) { IntArray(size) }
        for (x in 0 until size) {
            cells[x] = IntArray(size)
            for (y in 0 until size) {
                while (cells[x][y] == 0) {
                    cells[x][y] = Rnd.rnd(-weight, weight)
                }
            }
        }
        this.size = size
        this.weight = weight
        left = size * size
        caretX = size / 2
        caretY = size / 2
    }

    fun isCaretAt(x: Int, y: Int): Boolean {
        return x == caretX && y == caretY
    }

    fun isOpenBy(direction: Direction, step: Int): Boolean {
        return isOpenBy(direction.deltaX * step, direction.deltaY * step)
    }

    fun isOpenBy(move: Move): Boolean {
        return if (move == Move.OPEN) isOpenAtCaret else isOpenBy(requireNotNull(move.direction))
    }

    private val isOpenAtCaret: Boolean
        get() = cells[caretX][caretY] == 0

    private fun isOpenBy(direction: Direction): Boolean {
        return isOpenBy(direction.deltaX, direction.deltaY)
    }

    fun moveBy(move: Move) {
        if (move == Move.OPEN) {
            openAtCaret()
        } else {
            moveBy(requireNotNull(move.direction))
        }
    }

    fun openAtCaret(): Int {
        val r = cells[caretX][caretY]
        cells[caretX][caretY] = 0
        if (r != 0) {
            left--
        }
        return r
    }

    private fun moveBy(deltaX: Int, deltaY: Int): Boolean {
        val move = canMoveBy(deltaX, deltaY)
        if (move) {
            caretX += deltaX
            caretY += deltaY
        }
        return move
    }

    private fun moveBy(direction: Direction) {
        moveBy(direction.deltaX, direction.deltaY)
    }

    fun openBoard() {
        each { _: Int?, _: Int? -> 0 }
        left = 0
    }

    fun reInit() {
        init(size, weight)
    }

    fun moveBy(direction: Direction, step: Int): Boolean {
        return moveBy(direction.deltaX * step, direction.deltaY * step)
    }

    companion object {
        private val NO_CELLS = Array(0) { IntArray(0) }
    }
}
